File: apps/ios/ExpUI/Sources/GlassPill.swift
import SwiftUI

// The ONE pill (EXP-698 round 2). Before this file iOS had three pill types
// (`GlassPillLabel`, `GlassPillButton`, `GlassChip`) plus nineteen raw
// `.glassButton()` capsules, each picking its own font, padding and vertical
// rhythm — a filter pill, a property chip and a repo chip were three different
// heights sitting in the same row. `GlassPill` is the whole vocabulary:
//
//   size    .sm (24pt) — the DEFAULT, and what almost every pill is: a
//                         metadata chip, a filter, a property, a role badge,
//                         an inline action. A short pill beside a taller
//                         control is correct — the members list's role badge
//                         sits 12pt from the name and 12pt from the row's
//                         32pt circle button, and reads as a label, not as a
//                         second button.
//           .md (32pt) — reserved for a pill that is one of SEVERAL PEER
//                        ACTIONS in a bar of its own (the steer screen's
//                        Merge / Fix conflicts beside the Latest-changes
//                        chip, an empty state's primary action), where a
//                        24pt capsule reads as a caption instead of a
//                        button.
//   mode    .action    — a plain tap
//           .select    — a tap that also carries a selected state
//           .readonly  — a label capsule; not a Button, not hit-testable
//
// Chrome is the same in every combination: a `Capsule` wearing `fillCard` +
// the `strokeCard` hairline (`.glassButton()`), label white at
// `TextOpacity.secondary`. Only `.select` when SELECTED changes it, to
// `fillActive` + `strokeActive` with a full-opacity label — the same "selected"
// treatment a `GlassRow` and a segmented segment wear.
//
// `primary:` (EXP-698 r4) is PAINT, orthogonal to size and mode: the solid
// `primary` fill with `primaryForeground` content and no hairline, dimmed while
// pressed — the one loud pill in a view, for the action the screen exists for
// (issue detail's Watch). It is the same flag on all four clients (web
// `primary`, desktop `.primary()`, Android `primary =`), so a pill is never
// promoted by hand-painting a background at a call site. Disabled keeps the
// shared glass treatment: a disabled pill must not read as the loud one.

/// The pinned geometry of a glass pill, per size rung. Numbers live here
/// rather than inline so `GlassPillTokenTests` can hold them still: a pill
/// that quietly grows two points stops lining up with the circle buttons and
/// chips it shares a row with.
public enum GlassPillTokens {
    /// `controlSm` — the chip/filter rung.
    public static let heightSm: CGFloat = DesignTokens.Size.controlSm
    /// `controlMd` — the rung that matches `CircleIconButton`.
    public static let heightMd: CGFloat = DesignTokens.Size.controlMd

    public static let horizontalPaddingSm: CGFloat = 8
    public static let horizontalPaddingMd: CGFloat = 12

    /// Gap between the leading mark, the label and the trailing slot.
    public static let spacingSm: CGFloat = 4
    public static let spacingMd: CGFloat = 6

    /// The leading registry glyph.
    public static let glyphSm: CGFloat = 12
    public static let glyphMd: CGFloat = 16

    /// The status/label dot — a plain filled disc, never a ring.
    public static let dotSize: CGFloat = 6
}

/// The two rungs. A pill is either chip-sized or control-sized; there is no
/// third height and no free `verticalPadding` knob (the one EXP-678 added is
/// gone — the tall "Latest changes" pills are `.md`).
public enum GlassPillSize {
    case sm
    case md

    var height: CGFloat {
        self == .sm ? GlassPillTokens.heightSm : GlassPillTokens.heightMd
    }

    var horizontalPadding: CGFloat {
        self == .sm ? GlassPillTokens.horizontalPaddingSm : GlassPillTokens.horizontalPaddingMd
    }

    var spacing: CGFloat {
        self == .sm ? GlassPillTokens.spacingSm : GlassPillTokens.spacingMd
    }

    /// The leading glyph size — public so a call site building its own leading
    /// view (a brand image, a board icon, a spinner) matches the registry
    /// glyphs the `icon:` init draws.
    public var glyphSize: CGFloat {
        self == .sm ? GlassPillTokens.glyphSm : GlassPillTokens.glyphMd
    }

    var font: Font {
        self == .sm ? .caption.weight(.medium) : .subheadline.weight(.medium)
    }
}

public enum GlassPillMode {
    /// A plain tap.
    case action(() -> Void)
    /// A tap that also carries a selected state (a label toggle, a tab).
    case select(isSelected: Bool, action: () -> Void)
    /// A label capsule — resting chrome, no Button, no hit testing, so it never
    /// swallows a tap meant for the row behind it.
    case readonly
}

public struct GlassPill<Leading: View, Trailing: View>: View {

    /// Spelled out at call sites as `.sm` / `.md`; the type itself lives
    /// OUTSIDE the generic struct so every specialization shares one `Size`.
    public typealias Size = GlassPillSize
    public typealias Mode = GlassPillMode

    let label: String
    var size: GlassPillSize = .sm
    var mode: GlassPillMode = .readonly
    var dot: Color? = nil
    var isOpaque: Bool = false
    /// The loud paint (see the file header): solid `primary`, no hairline.
    var primary: Bool = false
    var enabled: Bool = true
    let leading: Leading
    let trailing: Trailing

    public init(
        _ label: String,
        size: GlassPillSize = .sm,
        mode: GlassPillMode = .readonly,
        dot: Color? = nil,
        isOpaque: Bool = false,
        primary: Bool = false,
        enabled: Bool = true,
        @ViewBuilder leading: () -> Leading,
        @ViewBuilder trailing: () -> Trailing
    ) {
        self.label = label
        self.size = size
        self.mode = mode
        self.dot = dot
        self.isOpaque = isOpaque
        self.primary = primary
        self.enabled = enabled
        self.leading = leading()
        self.trailing = trailing()
    }

    private var isSelected: Bool {
        if case .select(let selected, _) = mode { return selected }
        return false
    }

    /// A disabled pill is never the loud one — it keeps the shared glass
    /// treatment, which is what "disabled" looks like everywhere else.
    private var isPrimary: Bool { primary && enabled }

    private var labelOpacity: Double {
        guard enabled else { return TextOpacity.quaternary }
        return isSelected ? TextOpacity.primary : TextOpacity.secondary
    }

    @ViewBuilder
    public var body: some View {
        switch mode {
        case .readonly:
            content.allowsHitTesting(false)
        case .action(let action), .select(_, let action):
            if isPrimary {
                Button(action: action) {
                    content.contentShape(Capsule())
                }
                .buttonStyle(GlassPillPrimaryPressStyle())
                .disabled(!enabled)
            } else {
                Button(action: action) {
                    content.contentShape(Capsule())
                }
                .buttonStyle(.plain)
                .disabled(!enabled)
            }
        }
    }

    private var content: some View {
        HStack(spacing: size.spacing) {
            if let dot {
                Circle()
                    .fill(dot)
                    .frame(width: GlassPillTokens.dotSize, height: GlassPillTokens.dotSize)
            }
            leading
            if !label.isEmpty {
                Text(label)
                    .font(size.font)
                    .lineLimit(1)
            }
            trailing
        }
        .foregroundStyle(
            isPrimary
                ? DesignTokens.Palette.primaryForeground
                : Color.white.opacity(labelOpacity)
        )
        .padding(.horizontal, size.horizontalPadding)
        .frame(height: size.height)
        .modifier(Paint(isPrimary: isPrimary, isSelected: isSelected, isOpaque: isOpaque))
    }

    /// Glass or solid — the ONE branch between the two paints, so everything
    /// above it (geometry, font, label opacity) is shared.
    private struct Paint: ViewModifier {
        let isPrimary: Bool
        let isSelected: Bool
        let isOpaque: Bool

        func body(content: Content) -> some View {
            if isPrimary {
                content
                    .background(DesignTokens.Palette.primary, in: Capsule())
            } else {
                content
                    .glassButton(isActive: isSelected, isOpaque: isOpaque)
            }
        }
    }

}

/// The press feedback of a `primary` pill: a solid fill has no hairline to
/// brighten, so the fill itself dims — the same cue the submit button gives.
/// Public because a primary pill is not always its own Button: wrapped in a
/// `NavigationLink` (issue detail's Watch) the LINK owns the press, and
/// `.buttonStyle(.plain)` there would leave the loud pill inert under a finger.
public struct GlassPillPrimaryPressStyle: ButtonStyle {
    public init() {}

    public func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .opacity(configuration.isPressed ? 0.8 : 1)
    }
}

extension ButtonStyle where Self == GlassPillPrimaryPressStyle {
    /// `.buttonStyle(.glassPillPrimary)` — for the host that owns the tap of a
    /// `primary` pill it did not build as a Button.
    public static var glassPillPrimary: GlassPillPrimaryPressStyle {
        GlassPillPrimaryPressStyle()
    }
}

// MARK: - Convenience inits

extension GlassPill where Trailing == EmptyView {
    public init(
        _ label: String,
        size: GlassPillSize = .sm,
        mode: GlassPillMode = .readonly,
        dot: Color? = nil,
        isOpaque: Bool = false,
        primary: Bool = false,
        enabled: Bool = true,
        @ViewBuilder leading: () -> Leading
    ) {
        self.init(
            label,
            size: size,
            mode: mode,
            dot: dot,
            isOpaque: isOpaque,
            primary: primary,
            enabled: enabled,
            leading: leading
        ) { EmptyView() }
    }
}

extension GlassPill where Leading == EmptyView, Trailing == EmptyView {
    public init(
        _ label: String,
        size: GlassPillSize = .sm,
        mode: GlassPillMode = .readonly,
        dot: Color? = nil,
        isOpaque: Bool = false,
        primary: Bool = false,
        enabled: Bool = true
    ) {
        self.init(
            label,
            size: size,
            mode: mode,
            dot: dot,
            isOpaque: isOpaque,
            primary: primary,
            enabled: enabled
        ) { EmptyView() } trailing: { EmptyView() }
    }
}

extension GlassPill where Leading == AppIcon, Trailing == EmptyView {
    /// The common case: a registry glyph leading the label, sized to the rung.
    public init(
        _ label: String,
        icon: String,
        size: GlassPillSize = .sm,
        mode: GlassPillMode = .readonly,
        dot: Color? = nil,
        isOpaque: Bool = false,
        primary: Bool = false,
        enabled: Bool = true
    ) {
        self.init(
            label,
            size: size,
            mode: mode,
            dot: dot,
            isOpaque: isOpaque,
            primary: primary,
            enabled: enabled
        ) {
            AppIcon(icon, size: size.glyphSize)
        } trailing: { EmptyView() }
    }
}

File: apps/android/app/src/main/java/com/exponential/app/ui/components/GlassPill.kt
package com.exponential.app.ui.components

import androidx.compose.foundation.LocalIndication
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.exponential.app.ui.theme.DesignTokens
import com.exponential.app.ui.theme.GlassTokens
import com.exponential.app.ui.theme.TextEmphasis
import com.exponential.app.ui.theme.glassButton

/** The two pill rungs — a 32dp control and a 24dp inline badge. */
enum class PillSize { Md, Sm }

/**
 * What a pill IS, which is the only thing that may change its chrome:
 * [Action] taps and does something, [Select] carries a [GlassPill] `selected`
 * state, [Readonly] is a label that happens to be capsule-shaped.
 */
enum class PillMode { Action, Select, Readonly }

/**
 * The ONE capsule on Android (EXP-698).
 *
 * Before this the app drew fourteen of them: `GlassPillButton`, `PropertyChip`,
 * `FilterPill`, `LabelChip`, `OriginChip`, `RepoNameChip`, `SuggestionKindChip`,
 * `NoticeChip`, `AgentTabChip`, the members role badge, the sync chips, plus
 * twenty-five bare `Modifier.glassButton()` rows — each with its own height,
 * padding, glyph size and label alpha, and one with a hand-typed `0.9f`
 * emphasis that exists nowhere else in the design system. They are all this
 * now; [GlassPillDefaults] pins the numbers and `GlassPillDefaultsTest` keeps
 * them from being re-typed at a call site.
 *
 * Chrome is a pure function of ([size], [mode], [selected], [primary]): the
 * capsule is always [GlassTokens.CardFill] behind a [GlassTokens.StrokeCard] hairline
 * with a secondary-emphasis label, and ONLY a selected [PillMode.Select] pill
 * brightens to the shared active fill + stroke and a primary label. A
 * [PillMode.Readonly] pill takes the same rest chrome, is not clickable and
 * shows no ripple. [primary] is the one emphatic paint — a solid capsule for
 * the single call to action on a surface.
 *
 * [leading] is a free slot (a progress ring, a status glyph, an avatar) drawn
 * in the pill's own content color; [icon] is the common case of a plain glyph.
 * [dot] draws the 6dp colour disc a label pill needs (the label palette's
 * colour, the agent's brand tint). [trailing] is the dismiss `×` of a filter
 * pill.
 */
@Composable
fun GlassPill(
    /** Empty renders an icon-only capsule. */
    label: String,
    modifier: Modifier = Modifier,
    onClick: (() -> Unit)? = null,
    size: PillSize = PillSize.Md,
    mode: PillMode = if (onClick == null) PillMode.Readonly else PillMode.Action,
    selected: Boolean = false,
    /**
     * PAINT, orthogonal to [size] and [mode] (EXP-698 r4 — web `primary`,
     * desktop `.primary()`, iOS `primary:`): the solid primary fill with
     * primary-foreground content and no hairline, for the ONE call to action
     * on a surface. A disabled pill ignores it and keeps the dimmed glass.
     */
    primary: Boolean = false,
    icon: ImageVector? = null,
    leading: (@Composable () -> Unit)? = null,
    dot: Color? = null,
    trailing: (@Composable () -> Unit)? = null,
    opaque: Boolean = false,
    enabled: Boolean = true,
    /** A call in flight: the glyph slot becomes a spinner, so the pill keeps
     *  its width and the caller only has to dim it via [enabled]. */
    loading: Boolean = false,
    /** Monospace for an identifier or a `owner/repo` name (web `font-mono`,
     *  iOS `.monospaced()`). Null keeps the rung's own family. */
    fontFamily: FontFamily? = null,
    /**
     * ONE line, ellipsized — the only value a fixed-height capsule can honour,
     * so it is the default and raising it is a caller's explicit choice on a
     * pill it has also given room to grow. Text that can genuinely need two
     * lines is a NOTICE, not a pill (see the sync/steer captions).
     */
    maxLines: Int = 1,
    /** Overrides the emphasis-derived content color — the ONE destructive
     *  pill (Delete team) is red on the same glass. Nothing else may use it. */
    contentColor: Color? = null,
    /** Names the pill for accessibility when the LABEL does not — the
     *  icon-only capsule, whose glyph is its whole meaning. */
    contentDescription: String? = null,
) {
    // A DISABLED select pill is never lit: the active fill plus a quaternary
    // label rendered as a bright capsule with near-invisible text in it.
    val active = enabled && mode == PillMode.Select && selected
    // The emphatic paint is for a LIVE call to action only: a disabled primary
    // capsule would read as tappable at full brightness.
    val emphatic = enabled && primary
    val fg = contentColor ?: if (emphatic) {
        DesignTokens.Palette.PrimaryForeground
    } else {
        MaterialTheme.colorScheme.onSurface.copy(
            alpha = when {
                !enabled -> TextEmphasis.Quaternary
                active -> TextEmphasis.Primary
                else -> TextEmphasis.Secondary
            },
        )
    }
    val tap = if (enabled && mode != PillMode.Readonly) onClick else null
    // An icon-only capsule insets by whatever centres its glyph, so it comes
    // out a CIRCLE. The label padding would have made it a stubby 28x24 oval.
    val iconOnly = label.isEmpty()
    // Aliased because inside `semantics {}` the bare name resolves to THIS
    // parameter, not to the receiver's property of the same name.
    val describedAs = contentDescription
    // Only a solid fill needs a press state of its own — the glass rungs get
    // theirs from the ripple, and collecting the interaction stream for them
    // would recompose fifty capsules for nothing.
    val interactionSource = remember { MutableInteractionSource() }
    val pressed = if (emphatic) interactionSource.collectIsPressedAsState().value else false
    Row(
        verticalAlignment = Alignment.CenterVertically,
        // Centred, so a pill told to fillMaxWidth (the one destructive
        // "Delete team" capsule) centres its content instead of hugging the
        // left edge; it is a no-op on a pill that wraps its content.
        horizontalArrangement = Arrangement.spacedBy(
            GlassPillDefaults.spacing(size),
            Alignment.CenterHorizontally,
        ),
        modifier = modifier
            .height(GlassPillDefaults.height(size))
            .glassButton(active = active, opaque = opaque, primary = emphatic, pressed = pressed)
            .then(
                if (tap != null) {
                    Modifier.clickable(
                        interactionSource = interactionSource,
                        // The primary capsule's press cue IS the fill dipping;
                        // a ripple over a solid fill only muddies it.
                        indication = if (emphatic) null else LocalIndication.current,
                        onClick = tap,
                    )
                } else {
                    Modifier
                },
            )
            .then(
                if (describedAs != null) {
                    Modifier.semantics { this.contentDescription = describedAs }
                } else {
                    Modifier
                },
            )
            .padding(
                horizontal = if (iconOnly) {
                    GlassPillDefaults.iconOnlyHorizontalPadding(size)
                } else {
                    GlassPillDefaults.horizontalPadding(size)
                },
            ),
    ) {
        val glyph = GlassPillDefaults.glyphSize(size)
        when {
            loading -> CircularProgressIndicator(
                modifier = Modifier.size(glyph),
                strokeWidth = 2.dp,
                color = fg,
            )
            leading != null -> CompositionLocalProvider(LocalContentColor provides fg) { leading() }
            icon != null -> Icon(icon, contentDescription = null, modifier = Modifier.size(glyph), tint = fg)
        }
        if (dot != null) {
            Box(
                Modifier
                    .size(GlassPillDefaults.DotSize)
                    .clip(CircleShape)
                    .background(dot, CircleShape),
            )
        }
        // An EMPTY label is the icon-only pill (the issue header's "edit
        // properties" `+`) — it keeps the capsule, drops the text and the
        // spacing that would otherwise sit beside nothing.
        if (!iconOnly) {
            Text(
                label,
                style = GlassPillDefaults.textStyle(size).let {
                    if (fontFamily != null) it.copy(fontFamily = fontFamily) else it
                },
                color = fg,
                maxLines = maxLines,
                overflow = TextOverflow.Ellipsis,
            )
        }
        if (trailing != null) CompositionLocalProvider(LocalContentColor provides fg) { trailing() }
    }
}

/**
 * The pill's own numbers (EXP-698), pinned by `GlassPillDefaultsTest` — the
 * capsule is drawn at fifty call sites and none of them may re-type a height,
 * a padding or a glyph size.
 */
object GlassPillDefaults {
    /** [PillSize.Md] — the control rung every other 32dp control sits on. */
    val MdHeight: Dp = GlassTokens.ControlSize
    val MdHorizontalPadding: Dp = 12.dp
    val MdSpacing: Dp = 6.dp
    val MdGlyphSize: Dp = 16.dp

    /** [PillSize.Sm] — an inline badge inside a list row or a title line. */
    val SmHeight: Dp = 24.dp
    val SmHorizontalPadding: Dp = 8.dp
    val SmSpacing: Dp = 4.dp
    val SmGlyphSize: Dp = 12.dp

    /** The colour disc of a label / agent pill. */
    val DotSize: Dp = 6.dp

    /**
     * A person in a pill (the assignee chip). An avatar is a face, not a
     * glyph: at the 12dp glyph rung its initials fell to ~5sp and stopped
     * being readable, so it gets its own slightly larger rung.
     */
    val AvatarSize: Dp = 18.dp

    fun height(size: PillSize): Dp = if (size == PillSize.Md) MdHeight else SmHeight

    fun horizontalPadding(size: PillSize): Dp =
        if (size == PillSize.Md) MdHorizontalPadding else SmHorizontalPadding

    fun spacing(size: PillSize): Dp = if (size == PillSize.Md) MdSpacing else SmSpacing

    fun glyphSize(size: PillSize): Dp = if (size == PillSize.Md) MdGlyphSize else SmGlyphSize

    /** Whatever centres the glyph — an icon-only pill is a circle, not an oval. */
    fun iconOnlyHorizontalPadding(size: PillSize): Dp = (height(size) - glyphSize(size)) / 2

    /** Medium weight on both rungs — a pill never announces itself by weight. */
    @Composable
    fun textStyle(size: PillSize): TextStyle = when (size) {
        PillSize.Md -> MaterialTheme.typography.labelLarge
        PillSize.Sm -> MaterialTheme.typography.labelMedium
    }.copy(fontWeight = FontWeight.Medium)
}
